core/trino-main/src/test/java/io/trino/server/TestQueryResource.java

```diff
@@ -84,6 +84,40 @@ public void teardown()
         closeAll(server, client);
     }
 
+    @Test
+    public void testIdempotentResults()
+    {
+        String sql = "SELECT * FROM tpch.tiny.lineitem";
+
+        Request request = preparePost()
+                .setHeader(TRINO_HEADERS.requestUser(), "user")
+                .setUri(uriBuilderFrom(server.getBaseUrl().resolve("/v1/statement")).build())
+                .setBodyGenerator(createStaticBodyGenerator(sql, UTF_8))
+                .build();
+
+        QueryResults queryResults = client.execute(request, createJsonResponseHandler(jsonCodec(QueryResults.class)));
+        URI uri = queryResults.getNextUri();
+        while (uri != null) {
+            QueryResults attempt1 = client.execute(
+                    prepareGet()
+                            .setHeader(TRINO_HEADERS.requestUser(), "user")
+                            .setUri(uri)
+                            .build(),
+                    createJsonResponseHandler(jsonCodec(QueryResults.class)));
+
+            QueryResults attempt2 = client.execute(
+                    prepareGet()
+                            .setHeader(TRINO_HEADERS.requestUser(), "user")
+                            .setUri(uri)
+                            .build(),
+                    createJsonResponseHandler(jsonCodec(QueryResults.class)));
+
+            assertEquals(attempt2.getData(), attempt1.getData());
```

Comment:
We might want to verify the whole object here (via serializing to JSON) since the implementation caches the result. While in theory, we could update the stats between repeated requests, that isn't something we do today and seems unlikely to change.

Reply:
The stats do change in the initial phases of the protocol while nextUri points at the /queued resource, unfortunately.